docs/DeferredSignals.md

```diff
@@ -0,0 +1,246 @@
+* State reconstruction problems aren't discussed here.
+```
+! In JIT code
+! JIT code causes a synchronous signal (SIGSEGV or other)
+# Guest Signal Handler called
+# JIT jumps to guest signal handler
+# Hopefully guest calls rt_sigreturn instead of long jumping out.
+```
+
+### Asynchronous signal in code emitter
+This is the first interesting example since deferred signals affects it.
+```
+- <Enter Deferred region - 3 instructions>
+! Compiling JIT Code
+# Asynchronous Signal received
+# - Host signal handler determines the thread is in a deferred signal section.
+# - Signal information is stored in a queue
+# - mprotect signal page to NONE.
+# - Signal handler returns without giving the signal to the guest
+! Compiling JIT Code continues.
++ #1 - <Exit deferred region - 5 instructions>
++ Deferred region section causes SIGSEGV
++ - Host signal handler determines deferred region is done, Still has signal in queue.
++ - Pull signal information off of queue
+# JIT jumps to guest signal handler
+# Hopefully guest calls rt_sigreturn instead of long jumping out.
+# Host PC is back at deferred signal section.
++ Deferred region section causes SIGSEGV #2
++ - Host signal handler determines deferred region is done, No signals in the queue.
+# - mprotect signal page to RW.
+# - Continue execution.
++ #1 <Exit deferred region continues>
+```
+
+### Recursive regions with signal in code emitter.
+This one mostly matches the previous example except the behaviour of deferred signal regions leaving.
+
+In this case, if the thread-local refcount is still >0 on `<Exit deferred region>` then there are two behaviours.
+- On ARM64, it will receive a SIGSEGV but the signal handler will increment PC by one instruction and continue execution
+   - Expectation is that signals are significantly less common than `<Exit deferred region>` so the cost of SIGSEGV+PC increment is faster.
+- On x86-64, the region exit checks the refcount before doing the fault access.
+   - This adds more instructions so is slower on average.
+
+This has the expectation that recursive deferred regions both aren't very deep (usually only nested a couple times), and that signals are rare.
+This way there aren't many SIGSEGV checks generated and the signal is finally only handled when reaching the top-most deferred region exit routine.
+
+```
+- #1 <Enter Deferred region - 3 instructions>
+! Compiling JIT Code
+-    #2 <Enter Deferred region - 3 instructions>
+!    Memory allocation
+#    <Async signal received logic from above>
++    #2 <Exit deferred region - 5 instructions>
++    #2 Exit deferred region causes SIGSEGV
++    - TLS refcount is still 1 (from #1)
++    - PC is incremented by one instruction, signal still unhandled.
++ #1 <Exit deferred region - 5 instructions>
++ #1 Exit deferred region causes SIGSEGV
++ <Regular deferred region handling from above called>
+```
+
+### Multiple signals in deferred region
```

Comment:
"deferred region" isn't a good term since it suggests the region itself is somehow being deferred.

I think `s/deferred region/signal-deferring region/g` would be well-descriptive, though.